### ISSUE TITLE
#6336 causes overlord console page loading to be extreme slow

### ISSUE THREAD

**Issue by QiuMM** [apache/incubator-druid#6726]
As discussed in [#6336 (comment)](https://github.com/apache/incubator-druid/pull/6336#issuecomment-446211138).

**Comment by vogievetsky** [comment]
Is this still an issue in the latest code?

**Comment by QiuMM** [comment]
Yes, it is. I have made some improvement in my druid code, I'll open a PR to patch this when I'm not busy.

**Comment by FaxianZhao** [comment]
I suggest use a Fine-Grained Lock to resolve this kind issue once and for all, but I haven't go through the whole picture.

**Comment by FaxianZhao** [comment]
I suggest use the Fine-Grained Lock resolve this kind issues. 
I think my previous draft code won't cause this, because I use `synchronized (lockResult.getTaskLock())` instead of `toolbox.getTaskLockbox().doInCriticalSection`, but it looks bad for committed code.

**Comment by jihoonson** [comment]
@QiuMM thanks. I'm looking forward to your PR.

@FaxianZhao yeah, fine-grained locking would mitigate this issue. But, synchronizing `taskLock` isn't enough to fix the issue of https://github.com/apache/incubator-druid/pull/6336, because `taskLock` is shared by the tasks of the same groupId but the tasks of different groupIds should be able to request segment allocation for the same dataSource and interval at the same time. Check out https://github.com/apache/incubator-druid/issues/7491.

I think the best way to handle this is synchronizing segmentId allocation per dataSource per interval as @QiuMM did [here](https://github.com/apache/incubator-druid/pull/6336#issuecomment-450610474).
